homeassistant/components/minecraft_server/sensor.py

```diff
@@ -54,134 +135,28 @@ async def async_setup_entry(
 class MinecraftServerSensorEntity(MinecraftServerEntity, SensorEntity):
     """Representation of a Minecraft Server sensor base entity."""
 
+    entity_description: MinecraftServerSensorEntityDescription
+
     def __init__(
         self,
         server: MinecraftServer,
-        entity_type: str,
-        icon: str,
-        unit: str | None = None,
-        device_class: str | None = None,
+        description: MinecraftServerSensorEntityDescription,
     ) -> None:
         """Initialize sensor base entity."""
-        super().__init__(server, entity_type, icon, device_class)
-        self._attr_native_unit_of_measurement = unit
+        super().__init__(server)
+        self.entity_description = description
+        self._attr_unique_id = f"{server.unique_id}-{description.key}"
 
     @property
     def available(self) -> bool:
         """Return sensor availability."""
         return self._server.online
 
-
-class MinecraftServerVersionSensor(MinecraftServerSensorEntity):
-    """Representation of a Minecraft Server version sensor."""
-
-    _attr_translation_key = KEY_VERSION
-
-    def __init__(self, server: MinecraftServer) -> None:
-        """Initialize version sensor."""
-        super().__init__(server=server, entity_type=KEY_VERSION, icon=ICON_VERSION)
-
-    async def async_update(self) -> None:
-        """Update version."""
-        self._attr_native_value = self._server.data.version
-
-
-class MinecraftServerProtocolVersionSensor(MinecraftServerSensorEntity):
-    """Representation of a Minecraft Server protocol version sensor."""
-
-    _attr_translation_key = KEY_PROTOCOL_VERSION
-
-    def __init__(self, server: MinecraftServer) -> None:
-        """Initialize protocol version sensor."""
-        super().__init__(
-            server=server,
-            entity_type=KEY_PROTOCOL_VERSION,
-            icon=ICON_PROTOCOL_VERSION,
-        )
-
-    async def async_update(self) -> None:
-        """Update protocol version."""
-        self._attr_native_value = self._server.data.protocol_version
-
-
-class MinecraftServerLatencySensor(MinecraftServerSensorEntity):
-    """Representation of a Minecraft Server latency sensor."""
-
-    _attr_translation_key = KEY_LATENCY
-
-    def __init__(self, server: MinecraftServer) -> None:
-        """Initialize latency sensor."""
-        super().__init__(
-            server=server,
-            entity_type=KEY_LATENCY,
-            icon=ICON_LATENCY,
-            unit=UnitOfTime.MILLISECONDS,
-        )
-
     async def async_update(self) -> None:
```

Comment:
I am currently having some trouble understanding when this gets called. Does the `_server` update its own data? Do we really need to use a call to `async_update`?

I would prefer a setup where native value and the extra state attributes are provided using properties

Reply:
Exactly. The class `MinecraftServer` uses periodic updates and informs the platforms with a dispatcher signal to update the data via `async_update()`. It's basically similar to the data update coordinator. What you propose can and should probably be done while switching to the data update coordinator. But for now I fear we have to keep it like it is.